pkg/lib/cockpit-components-terminal.jsx

```diff
@@ -120,70 +196,74 @@ export class Terminal extends React.Component {
-                 * it to the websocket would result in a single 0xA0, which is invalid UTF-8 (and causes the session
-                 * to crash). So intercept and ignore such broken chars.
-                 * See https://github.com/cockpit-project/cockpit/issues/21213 */
-                if (data === '\x00') {
-                    console.log("terminal: ignoring invalid input", data);
-                    return;
-                }
-                this.props.channel.send(data);
-            }
-        }.bind(this));
-
         if (props.onTitleChanged)
             term.onTitleChange(props.onTitleChanged);
 
         this.terminal = term;
         this.state = {
             showPastingModal: false,
-            cols: props.cols || 80,
-            rows: props.rows || 25
+            cols: term.cols,
+            rows: term.rows
         };
     }
 
-    componentDidMount() {
-        this.terminal.open(this.terminalRef.current);
-        this.terminal.loadAddon(new CanvasAddon());
-        this.connectChannel();
+    mountTerminal(state) {
+        this.terminal = state.terminal;
+        this.terminalRef.current.appendChild(state.element);
+        this.terminal.open(state.element);
+        state.connectChannel(this.props.channel);
+
+        if (this.props.fontSize)
+            this.terminal.options.fontSize = this.props.fontSize;
+
+        if (this.props.rows) {
+            this.resizeTerminal(this.props.cols, this.props.rows);
+        }
 
+        this.setTerminalTheme(this.props.theme || 'black-theme');
+        this.terminal.focus();
+    }
+
+    unmountTerminal(state) {
+        this.terminalRef.current.removeChild(state.element);
+        if (!this.props.rows)
+            this.resizeTerminal(80, 1);
+    }
+
+    componentDidMount() {
+        this.mountTerminal(this.terminal_state);
         if (!this.props.rows) {
             window.addEventListener('resize', this.onWindowResize);
             this.onWindowResize();
         }
-        this.setTerminalTheme(this.props.theme || 'black-theme');
-        this.terminal.focus();
     }
 
     resizeTerminal(cols, rows) {
         this.terminal.resize(cols, rows);
-        this.props.channel.control({
-            window: {
-                rows,
-                cols
-            }
-        });
+        if (this.terminal_state.channel) {
+            this.terminal_state.channel.control({
+                window: {
+                    rows,
+                    cols
+                }
+            });
+        }
     }
 
     componentDidUpdate(prevProps, prevState) {
+        if (prevProps.state !== this.props.state) {
+            this.unmountTerminal(prevProps.state);
+            this.mountTerminal(this.props.state);
+            this.terminal_state = this.props.state;
+            if (!this.props.rows)
+                this.resizeTerminal(this.state.cols, this.state.rows);
```

Comment:
These 6 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-6bb6065b-20250807-121948-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)